src/dune_rules/melange_rules.ml

```diff
@@ -9,6 +9,28 @@ let make_js_name ~js_ext ~dst_dir m =
   let name = Melange.js_basename m ^ js_ext in
   Path.Build.relative dst_dir name
 
+let virtual_dst_dir_and_modules sctx ~target_dir ~implements =
+  let open Memo.O in
+  match implements with
+  | None -> Memo.return None
+  | Some vlib ->
+    let* vlib = vlib in
+    let* vlib = Resolve.read_memo vlib in
+    let name = Lib.name vlib in
+    let vlib = Lib.Local.of_lib_exn vlib in
+    let info = Lib.Local.info vlib in
+    let* dir_contents =
+      let dir = Lib_info.src_dir info in
+      Dir_contents.get sctx ~dir
+    in
+    let* modules =
+      Dir_contents.ocaml dir_contents
+      >>| Ml_sources.modules ~for_:(Library name)
+    in
+    let lib_dir = Lib_info.src_dir info in
+    let dst_dir = lib_output_dir ~target_dir ~lib_dir in
+    Memo.return (Some (dst_dir, modules))
```

Comment:
There's Dir_contents.modules_of_lib to simplify this.

Reply:
https://github.com/ocaml/dune/pull/6495/commits/f9b9ab21f2c82ddfd533fd0ef0cbdc1514209884